enhancements/cluster-logging/fluentd-tuning.md

```diff
@@ -0,0 +1,146 @@
+---
+title: cluster-logging-log-forwarding
+authors:
+  - "@alanconway"
+reviewers:
+approvers:
+creation-date: 2020-05-27
+status: provisional
+see-also:[]
+replaces:[]
+superseded-by:[]
+---
+
+# Fluentd Peformance Tuning
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+Allow defaults for selected fluentd `<buffer>` fields to be specified globally
+for logging, for performance-tuning purposes.
+
+Tuning settings are:
+* *not* relevant to most users, default settings should give good general performance.
+* only for *advanced* users with detailed knowledge of `fluentd` configuration and performance.
+* only for performance tuning, they have no effect on functional aspects of logging.
+
+Tuning settings have weaker forward-compatibility guarantees than the rest of the API.
+In future releases:
+  - `fluentd` tuning may be ignored, if the logging implementation changes.
+  - values set in tuning may be overridden by settings in other logging APIs.
+  - new tuning sections may be introduced for new implementations
+    - they may not have equivalent settings to `tunings.fluetnd`
+
+Tuning will be presented in the `clusterloggings.logging.openshift.io` API as a
```

Comment:
> We currently have a collector(tion?).fluent section. Do we really want to introduce a a new "parent" in lieu of placing it under their? 

My concern is these parameters are not tuning *collection*, they are tuning *forwarding*. Buffer sizes, flush intervals, retries all describe how to *send* logs (forwarding), they don't describe which logs we collect or from where (collection). 
 
We happen to use fluentd as both collector and forwarder today, which creates confusion over the boundary between collection and forwarding. I'm wary of adding to that confusion by putting forwarder tuning into an API called "collector". It could work, and we could justify it on the basis that this is all just short term hackery, but I'd rather put things in their proper place even for the short term.

> Conceptually we could provide a mechanism by which you tune elasticsearch and wouldnt that make more sense to leave under the existing store?

Yes, but Elasticsearch *is* a store. Fluentd is both collector and forwarder, so I think the consistent choices are:
* split the config, input->collector API, output->forwarder API
* use clusterlogging API as a neutral dumping ground for global settings

Reply:
II haven't paid much attention to ClusterLogging, time to do so. IMO putting it in collection is architecturally wrong, and we'd be creating public-API tech debt, worst kind.  If we are soon replacing ClusterLogging that is to our advantage - we will have a chance to review and maybe improve this thing (which is basically an ugly hack) or carry it forward if we can't think of anything better. Let me review the forwarder option again and come back with some concrete proposals and examples before we decide.